### PR TITLE
[Merged by Bors] - refactor: do not allow `qsmul` to default automatically

### PR DESCRIPTION
**Pull request by eric-wieser**
Follows on from #6262.
Again, this does not attempt to fix any diamonds; it only identifies where they may be.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

File: Mathlib/Algebra/Order/CauSeq/Completion.lean [hunk 1]
```diff
@@ -276,6 +276,7 @@ noncomputable instance Cauchy.divisionRing : DivisionRing (Cauchy abv) where
   mul_inv_cancel x := CauSeq.Completion.mul_inv_cancel
   ratCast q := ofRat q
   ratCast_mk n d hd hnd := by rw [← ofRat_ratCast, Rat.cast_mk', ofRat_mul, ofRat_inv]; rfl
+  qsmul := qsmulRec _ -- TODO: fix instance diamond
```
Comment (mattrobball) [inline Mathlib/Algebra/Order/CauSeq/Completion.lean:279]: Maybe make a separate issue or add to an existing one.

**Review by mattrobball — APPROVED** [review]

File: Mathlib/RingTheory/IntegralDomain.lean [hunk 2]
```diff
@@ -96,13 +96,14 @@ variable [Ring R] [IsDomain R] [Fintype R]
...
-  { show GroupWithZero R from Fintype.groupWithZeroOfCancel R, ‹Ring R› with }
+  { show GroupWithZero R from Fintype.groupWithZeroOfCancel R, ‹Ring R› with
+    qsmul := qsmulRec _}
 #align fintype.division_ring_of_is_domain Fintype.divisionRingOfIsDomain
 
 /-- Every finite commutative domain is a field. More generally, commutativity is not required: this
 can be found in `Mathlib.RingTheory.LittleWedderburn`. -/
 def Fintype.fieldOfDomain (R) [CommRing R] [IsDomain R] [DecidableEq R] [Fintype R] : Field R :=
-  { Fintype.groupWithZeroOfCancel R, ‹CommRing R› with }
+  { Fintype.divisionRingOfIsDomain R, ‹CommRing R› with }
```
Comment (eric-wieser) [inline Mathlib/RingTheory/IntegralDomain.lean:106]: I guess this caused the speedup?

File: Mathlib/Algebra/Order/CauSeq/Completion.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Algebra/Order/CauSeq/Completion.lean:279]: Easier to just fix it, #11263

**Comment by mattrobball** [comment]
bors merge

**Comment by YaelDillies** [comment]
Is that PR the start of the doom of Type-valued default fields?

**Comment by eric-wieser** [comment]
For now I think it's only the ones involved in diamonds that matter. There'd be no benefit to changing `sub` and `div`.

**Comment by eric-wieser** [comment]
`pow` and `zpow` are probably also not worth the trouble; but `nat/int/ratCast` are next in line.

**Pull request closed without merging**